crates/metadata-store/src/rocksdb/server.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use bytes::Bytes;
+use bytestring::ByteString;
+use restate_core::cancellation_watcher;
+use restate_core::network::{MessageHandler, NetworkSender};
+use restate_node_protocol::metadata_store::{
+    Delete, Get, GetVersion, MetadataStoreRequest, MetadataStoreResponse, Put, RequestKind,
+    ResponseKind,
+};
+use restate_node_protocol::{metadata_store, MessageEnvelope};
+use restate_types::errors::GenericError;
+use restate_types::metadata_store::{Precondition, VersionedValue};
+use restate_types::{GenerationalNodeId, NodeId, Version};
+use rocksdb::{ColumnFamily, Options, WriteOptions, DB};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::path::Path;
+use tokio::sync::mpsc;
+use tracing::{debug, trace};
+
+// todo: Introduce bound if this channel should grow unboundedly
+type CommandSender = mpsc::UnboundedSender<(GenerationalNodeId, MetadataStoreRequest)>;
+type CommandReceiver = mpsc::UnboundedReceiver<(GenerationalNodeId, MetadataStoreRequest)>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("storage error: {0}")]
+    Storage(#[from] rocksdb::Error),
+    #[error("violated precondition: {0}")]
+    PreconditionViolation(String),
+    #[error("codec error: {0}")]
+    Codec(GenericError),
+}
```

Comment:
I guess so. Will try how it looks with this change.

Reply:
It does not seem to save a lot because now I need to add some boilerplate code which converts from `rocksdb::Error` into `metadata_store::Error::Internal`.